xdsl/irdl/declarative_assembly_format.py

```diff
@@ -119,6 +134,32 @@ def parse(
             result_types=result_types, operands=operands, attributes=state.attributes
         )
 
+    def resolve_operand_types(self, state: ParsingState) -> None:
+        for i, operand_type in enumerate(state.operand_types):
+            if operand_type is None:
+                state.operand_types[i] = self._resolve_type(
+                    state, VarIRConstruct.OPERAND, i
+                )
+
+    def resolve_result_types(self, state: ParsingState) -> None:
+        for i, result_type in enumerate(state.result_types):
+            if result_type is None:
+                state.result_types[i] = self._resolve_type(
+                    state, VarIRConstruct.RESULT, i
+                )
+
+    def _resolve_type(self, state: ParsingState, construct: VarIRConstruct, index: int):
+        resolve, (construct, idx) = self.type_resolutions[construct, index]
+        match construct:
+            case VarIRConstruct.OPERAND:
+                input_type = state.operand_types[idx]
+            case VarIRConstruct.RESULT:
+                input_type = state.result_types[idx]
+            case _:
+                raise ValueError()
```

Comment:
Would be nice to add a string here and a test that reaches it

Reply:
Not done :+1: 
Reworked the typing to let pyright know that the two handle cases are the only possible ones :yum: